src/ucp/api/ucp.h

```diff
@@ -3311,13 +3318,19 @@ ucp_atomic_fetch_nb(ucp_ep_h ep, ucp_atomic_fetch_op_t opcode,
  * memory address @a remote_addr and the @ref ucp_rkey_h "remote memory handle"
  * @a rkey.
  * The routine is non-blocking and therefore returns immediately. However the
- * actual atomic operation may be delayed. The atomic operation is not considered complete
- * until the values in remote and local memory are completed.
- *
- * @note The user should not modify any part of the @a buffer or @a result after
- *       this operation is called, until the operation completes.
+ * actual atomic operation may be delayed. The atomic operation is not considered
+ * complete until the values in remote and local memory are completed. The
+ * result of the operation will be stored in the @a param->reply_buffer memory
+ * buffer. In the case of CSWAP the value in @a param->reply_buffer will be
+ * swapped into the @a remote_addr if the condition is true.
+ *
+ * @note The user should not modify any part of the @a buffer or
+ *       @a param->reply_buffer after this operation is called, until the
+ *       operation completes.
  * @note Only ucp_dt_make_config(4) and ucp_dt_make_contig(8) are supported in
  *       @a param->datatype, see @ref ucp_dt_make_contig
+ * @note This operation requires @a param->reply_buffer and its corresponding
+ *       flag @ref UCP_OP_ATTR_FIELD_REPLY_BUFFER to be defined in @a param.
```

Comment:
The most confusing part of this that this is not really optional parameter for fetching operation ? Fetching operations always will require a buffer to store the result, is not it ? In this case why it is not part of argument list but an optional parameter ?

Reply:
@brminich maybe we can say that if reply is not specified, the operation behaves like "post" and returns when the send buffer is reusable?

Reply:
@yosefe In this case this is not a fetch operation and probably should go as a separate API.

Reply:
@yosefe, but it is called `fetch` also we have `post` already
@shamisp, we plan to support reply_buf for AM API, where it will be optional, so we will add it to the  `ucp_request_param_t`. Thus, it would more consistent to reuse it here as well.
BTW, for this atomic call we already have mandatory field in params - it is datatype.

Reply:
So, this the field that always have to be defined and not optional. @brminich thanks for highlighting the datatype issue. From API design point, it looks very concerting.

BTW, I'm getting your point about AM API. Maybe the right way to do it, is to introduce union inside of param, where each union will represent different operation class.

Reply:
> BTW, I'm getting your point about AM API. Maybe the right way to do it, is to introduce union inside of param, where each union will represent different operation class.

Unfortunately, there is a problem with this: if we add the union "by value" than the content of the union will not be expandable, and if we add it "by pointer", it will be yet another pointer dereference on fast path

so instead of try to generalize the fields in request_param_t so the same field would be useful for more than one operation

Reply:
@yosefe Let's take this conversation offline ?

Reply:
@shamisp, btw `ucp_request_param_t` does not necessarily contain only optional parameters.
So, imo, better keep common for several APIs stuff inside params for consistency.
wdyt?

Reply:
@brminich - I know. This is design issue that I want to discuss with @yosefe.

Reply:
@shamisp, what format do you prefer? Can we arrange some meeting or ?

Reply:
@brminich - working on it.